java/src/jmri/jmrit/simpleclock/SimpleClockFrame.java

```diff
@@ -514,7 +514,7 @@ void updateRunningButton() {
                     Bundle.getMessage("ErrorTitle"), JOptionPane.ERROR_MESSAGE);
             return null;
         }
-        if (InstanceManager.getDefault(jmri.ClockControl.class).requiresIntegerRate()) {
+        if (InstanceManager.getDefault(jmri.ClockControl.class).requiresIntegerRate() && !clock.getInternalMaster()) {
```

Comment:
Okay, now I see where the confusion comes from - hope I'm using the github interface right to explain better.

This change here in line 517 is the change that actually allows setting non-integer rates (by skipping the check if time source is internal).

The other change down in setTimeSourceChanged which you reference avoids the following workflow:

1) Set Time Source to internal
2) Change rate to non-integer value
3) Set Time Source back to loconet

So step 3 is not allowed when the current rate is not an integer. That's why that check is in the else block.

(and I will separate such commits in the future to make this more clear)

Reply:
And apart from that, I believe your assumptions about internalMaster and internalSourceIndex are correct, but I also find the clock stuff pretty obscure, so I'm not 100% sure either...

Thanks for taking the time to write this up in so much detail.